mining: Add mining test harness.

rstaudt2 opened the pull request:
This adds a mining test harness that includes functionality for creating and signing transactions, adding/removing transactions to a fake tx source to consider for inclusion in new blocks, and a fake chain that provides a mocked chain state as well as utxos for use in generating valid transactions.

Additionally, it adds tests that provide some initial test coverage for new block template generation using the mining harness and moves the TxMiningView tests from mempool to mining.

---

A few notes on the implementation:
- The `fakeTxSource` type became a bit bigger than I originally thought it would need to be (borrowing much from the actual `mempool` implementation).  This is mainly because, in addition to providing a `TxSource`, it is handling orphan and stage transactions so that out of order insertion into the mining view can be properly tested.
- I did not make the `miningHarness` or underlying `fakeChain` and `fakeTxSource` instances safe for concurrent access.  It is set up so that each test should be creating a new mining harness via `newMiningHarness` so that as the test changes the state of the harness or mocks certain functions, it does not impact any other tests. Since this is the case, making it safe for concurrent access seemed unnecessary at this point.

---

Next steps:
- Now that the mining view tests are moved to `mining`, a small interface can be created for `TxMiningView` in `mempool`.  I'll follow up with this in a separate PR.
- Further test coverage can be added, as this is only covering some basic scenarios currently, and I expect that some iteration will be needed on the `miningHarness` as more coverage is added.
- With the test framework in place, we can now (more easily) start taking on some optimizations for the template generation code.

dnldd commented on internal/mining/mining_harness_test.go:
```diff
@@ -0,0 +1,1431 @@
+}
+
+// isTransactionInTxSource returns whether or not the passed transaction exists
+// in the main pool.
```
s / ... in the main pool / in the mempool.

dnldd commented on internal/mining/mining_harness_test.go:
```diff
@@ -0,0 +1,1431 @@
+			// discarded.
+			p.removeStagedTransaction(redeemer)
+			_, err := p.maybeAcceptTransaction(redeemer, true)
+
```
extra line.

dnldd reviewed (commented):
Looks great overall :+1:

rstaudt2 commented:
Thanks for the review, @dnldd.  I corrected the extra line you pointed out, and I also added an additional validation in `TestNewBlockTemplate` to call `blockchain.CheckBlockSanity` and ensure that the generated block template is sane.

davecgh reviewed (changes requested):
Really great job overall.  I have some review comments inline.

For reference, @rstaudt2 and I discussed that this could potentially be based on actually generating a full chain via `chaingen`, but mocking out the current best state is most likely sufficient since any conditions that test would realistically catch are very likely to be the ones they're specifically testing for anyway, in which case the end result of having a mock state to trigger them versus doing so via a fully generated chain would be the same.

Also, since it's mocked out, we can always extend it in that fashion in the future should it become necessary.

rstaudt2 commented on internal/mining/mining_harness_test.go:
```diff
@@ -0,0 +1,1437 @@
+// isTransactionInTxSource returns whether or not the passed transaction exists
+// in the main pool.
+func (p *fakeTxSource) isTransactionInTxSource(hash *chainhash.Hash) bool {
+	if _, exists := p.pool[*hash]; exists {
```
Agreed, simplified these in the test code.

rstaudt2 commented on internal/mining/mining_harness_test.go:
```diff
@@ -0,0 +1,1437 @@
+
+// removeOrphan removes the passed orphan transaction from the orphan pool.
+func (p *fakeTxSource) removeOrphan(tx *dcrutil.Tx, removeRedeemers bool, isTreasuryEnabled bool) {
+	// Nothing to do if passed tx is not an orphan.
```
Updated this to more explicitly match the logic and say `Nothing to do if the passed tx does not exist in the orphan pool.`

rstaudt2 commented on internal/mining/mining_test.go:
```diff
@@ -0,0 +1,189 @@
+		return 0, errors.New("error retrieving standard verify flags")
+	}
+	_, err = harness.generator.NewBlockTemplate(address)
+	if err == nil {
```
That's a great call.  I meant to come back to this to check for the specific errors, and the way you outlined it is a good way to handle it.  Updated.

rstaudt2 commented on internal/mining/mining_harness_test.go:
```diff
@@ -0,0 +1,1437 @@
+// AddFakeUTXO creates a fake mined utxo for the provided transaction.
+func (m *miningHarness) AddFakeUTXO(tx *dcrutil.Tx, blockHeight int64, blockIndex uint32) {
+	m.chain.utxos.AddTxOuts(tx, blockHeight, blockIndex,
+		m.chain.isTreasuryAgendaActive)
```
The idea was that the tests can set/unset `m.chain.isTreasuryAgendaActive`, `isHeaderCommitmentsAgendaActive`, etc. depending on what it is trying to simulate.

davecgh replied:
Right, I understand that part of having the variable in the harness for the general case and it both makes sense and is a good idea I think, but I was really referring more to the fact that each individual utxo necessarily comes from some other place than the current tip of whatever the harness is associated with and therefore does not necessarily match whatever the current mocked state of the agenda is for the chain.

This would mean the caller would end up having to do something like the following for every utxo being mocked from some historical standpoint:

```Go
curTrsyActive := harness.chain.isTreasuryAgendaActive
harness.AddFakeUTXO(tx, someHeight, someIndex)
harness.chain.isTreasuryAgendaActive = curTrsyActive
```

rstaudt2 replied:
True, given that it is passing in the height and index as parameters, it probably makes more sense to pass isTreasuryAgendaActive as a parameter as well.  I'll update it to that.

rstaudt2 replied:
Updated this and also rebased on master.

rstaudt2 commented on internal/mining/mining_harness_test.go:
```diff
@@ -0,0 +1,1437 @@
+	}
+
+	// Create a SubsidyCache instance.
+	subsidyCache := standalone.NewSubsidyCache(chainParams)
```
This was just created separately since the same `subsidyCache` instance is being passed to both `fakeTxSource` and `miningHarness` below.

davecgh replied:
Make sense.

rstaudt2 commented:
@davecgh Thanks for the review!  I believe all of your feedback is addressed, and I also added an additional commit to update the comments/stylistic things that also existed in `mempool`, as you noted.
